File: 2019-2020_programmers/_16_/src/java/SortedArrToBst.java
import java.util.Arrays;

public class SortedArrToBst {
    public static void main(String[] args) {
        int[] sortedArr = {-10, -3, 0, 5, 9};
        printTree(sortedArrayToBST(sortedArr));
    }
    public static TreeNode sortedArrayToBST(int[] arr) {
        if (arr.length == 0) return null;
        /*log/
        int mid;
        */
        TreeNode node = new TreeNode(arr[arr.length / 2]);
        /* log/
        System.out.println("length: " + sortedArr.length + ", array: " +  Arrays.stream(sortedArr).boxed().collect(Collectors.toCollection(LinkedList::new)));
        System.out.println("mid: " + (mid = sortedArr.length) / 2 + ", el: " + sortedArr[mid / 2]);
        System.out.println(node + "\n");
         */
        int[] left = Arrays.stream(arr).limit(arr.length / 2).toArray();
        int[] right = Arrays.stream(arr).skip(arr.length / 2 + 1).limit(arr.length / 2).toArray();
        node.left = sortedArrayToBST(left);
        node.right = sortedArrayToBST(right);
        return node;
    }

    public static void printTree(TreeNode node) {
        if (node == null) return;
        printTree(node.left);
        System.out.println(node.val);
        printTree(node.right);
    }
}


class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int val) { this.val = val; }
    @Override
    public String toString() {
        return "N(" + this.val + ")";
    }
}

File: 2019-2020_programmers/_16_/src/kotlin/SortedArrToBst.kt
import java.util.Arrays;
import java.util.LinkedList;
import java.util.stream.Collectors;

fun main(args: Array<String>) {
   val sortedArr: IntArray = intArrayOf(-10, -3, 0, 5, 9)
   SortedArrToBst().printTree(sortedArrayToBST(sortedArr));
}
public class SortedArrToBst {
   fun sortedArrayToBST(arr: IntArray): TreeNode? {
       if (arr.size == 0) return null;
       /*log/
       */
       var mid: Int;
       var node: TreeNode = TreeNode(arr[arr.size/ 2]);
       /* log/
        */
//        println("length: $arr.length , array: ${Arrays.stream(arr).boxed().collect(Collectors.toCollection(LinkedList::new)))}"
       mid = arr.size
       println("mid: ${mid / 2}, el: ${arr[mid / 2]}")
       println("$node\n");
       var left = Arrays.stream(arr).limit((arr.size / 2).toLong()).toArray();
       var right = Arrays.stream(arr).skip((arr.size/ 2 + 1).toLong()).limit((arr.size/ 2).toLong()).toArray();
       node.left = sortedArrayToBST(left);
       node.right = sortedArrayToBST(right);
       return node;
   }

  fun printTree(node: TreeNode): Unit {
       if (node == null) return
       printTree(node.left!!)
       println(node.value)
       printTree(node.right!!)
   }
}


class TreeNode(val value: Int) {
   var left: TreeNode? = null
   var right: TreeNode? = null
   override fun toString(): String  {
       return "Node($this.val)"
   }
}
